### PR TITLE
Salmon and Transcriptome Index Tests Documentation

### PR DESCRIPTION
**Pull request by bug-sam**

### CODE REVIEW

**Review by kurtwheeler — CHANGES_REQUESTED** [review]
I think overall this is pretty great!! However I think there's some changes you could make to make it a bit easier to digest.

What we're really looking to be able to answer is "What is being tested?", so I think focusing more on the functions that are actually running the tests and less on the helper functions would be a bit clearer. So I think the goal should be to outline what each test tests, at a fairly high level of detail. If you end up repeating some text for several tests, that's okay.

File: salmon.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+### `identical_checksum`
+
+Checks if two files have identical checksums.
+
+### `strong_quant_correlation`
+
+Checks if the two inputed quant files are strongly correlated.
+
+They are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a correlation >= 0.99.
```
Comment (kurtwheeler) [inline salmon.md]: This is a very useful level of detail, however it's a bit too far from the tests that actually use it.

File: salmon.md [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+- calls `check_salmon_quant`
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
+- asserts that the experiment is not ready for tximport yet - salmon qunat still needs to be run on sample2
+- asserts that there are no tximport outputs yet
+- calls `salmon._prepare_files()`, passing in  og_file_2
+- cleans up tximport data from previous runs if it exists
+- calls `check_salmon_quant`
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
```
Comment (kurtwheeler) [inline salmon.md:111]:
```suggestion
    - asserts quant output file strongly correlates to reference file. This is determined to be the case if both columns #3 and #4 (zero-indexed) of the two files have a correlation >= 0.99.
```

Including the detail about what's being tested here means we won't also have to describe it above, and then all the information a scientist might want to know is in one place.  (Although they may also want to know what correlation is being used, e.g., spearman, pearson)

File: salmon.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    - incomplete samples has 5 accession codes
+    - complete samples has 20 accessioncodes
+    - By having 20/25 samples complete, we're just past both the numerical and percent cutoffs for running tximport early
+- calls `run_tximport_at_progress_point` to build the `job_context`, create the experiment with samples, and run tximport
+- asserts that 'tximported' is in the final `job_context`
+- for each sample in the completed samples list:
+    - check that the associations were made properly in `create_tximport_job_context`
+    - check that the tpm and rds files were created for the sample
+    - check that the sample was processed
+- for each sample in the incoplete sample list:
```
Comment (kurtwheeler) [inline salmon.md]:
```suggestion
- for each sample in the incomplete sample list:
```

File: salmon.md [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+
+Tests that the correct index length is calculated when a sample has one OriginalFile (single read)
+
+test steps:
+- calls `prepare_job`
+- creates a dummy `job_context` with one OriginalFile
+- calls `salmon._prepare_files` to prepare the OriginalFile before determining the index length
+- calls `salmon._determine_index_length`
+- asserts that the raw index length is correctly set
+- asserts that the indexl length is correctly set
```
Comment (kurtwheeler) [inline salmon.md]:
```suggestion
- asserts that the index length is correctly set
```

File: salmon.md [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+Tests that the correct index length is calculated when a sample has two OriginalFiles (double read)
+
+test steps:
+- calls `prepare_job`
+- creates a dummy `job_context` with both OriginalFiles
+- calls `salmon._prepare_files` to prepare the OriginalFiles before determining the index length
+- calls `salmon._determine_index_length`
+- asserts that the raw index length is correctly set
+- asserts that the indexl length is correctly set
```
Comment (kurtwheeler) [inline salmon.md]:
```suggestion
- asserts that the index length is correctly set
```

**Comment by kurtwheeler** [comment]
I think it might also be nice to include links to the sub-readmes from the main README.md.

**Review by kurtwheeler — APPROVED** [review]
This looks good to me! Nice work!

You should see what feedback Jackie has for you if any though.

File: salmon.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+- asserts that the `job.success` is true
+- asserts that the `sample.is_processed` is false
+- asserts that the index type for the organism is 'TRANSCRIPTOME_SHORT'
+
+### `test_no_salmon_on_geo`
+
+Confirms that salmon won't be run on GEO data
+
+test steps:
+- calls `prepere_job`
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
- calls `prepare_job`
```

File: salmon.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+Tests the entire pipeline with a sra file
+
+test steps:
+- calls `prepare_dotsra_job`
+    - creates all prerequisites for the salmon pipeline to be run including:
+        - ProcessorJob
+        - Sample
+        - OriginalFiles
+        - OrganismIndex
+    - uses an RSA file when creating the OriginalFile
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
    - uses an SRA file when creating the OriginalFile
```

I think?

File: salmon.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+Confirms that the job fails with an SRA file that doesn't exist
+
+test steps:
+- calls `prepare_dotsra_job` with a file that doesn't exist
+    - creates all prerequisites for the salmon pipeline to be run including:
+        - ProcessorJob
+        - Sample
+        - OriginalFiles
+        - OrganismIndex
+    - uses an RSA file when creating the OriginalFile
```
Comment (jaclyn-taroni) [inline salmon.md]: Same comment as above
```suggestion
    - uses an SRA file when creating the OriginalFile
```

File: salmon.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+    - creates a dummy OrganismIndex that can be used in the salmon pipeline
+- creates the test sample and experiment
+- adds an additional fake sample to the experiment to prevent tximport from running on the experiment
+    - tximport will only run after salmon quant is run on every sample associated with the experiment
+- creates two original files and adds them to the test sample
+- calls `salmon._prepare_files` to prepare the original files before running salmon quant
+- calls `check_salmon_quant` to run salmon quant and tximport
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
+    - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.
```
Comment (jaclyn-taroni) [inline salmon.md]: Adding what those columns in the quant file are
```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

File: salmon.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        - OriginalFiles
+        - OrganismIndex
+    - uses an RSA file when creating the OriginalFile
+- passes the created base job to `salmon.salmon()` which runs the entire pipeline
+- asserts that `job_context` was created
+- asserts that the `job.success` is false
+
+### `test_salmon_quant_one_sample_double_reads`
+
+Tests salmon quant on a single sample with two original files (double reads)
```
Comment (jaclyn-taroni) [inline salmon.md]: I think this is what this means / how I would talk about it
```suggestion
Tests salmon quant on a single sample with two original files (paired end)
```

File: salmon.md [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+    - creates a dummy OrganismIndex that can be used in the salmon pipeline
+- creates the test experiment
+- creates sample1 and og_file_1, associates the two, and adds the sample to the experiemnt
+- creates sample2 and og_file_2, associates the two, and adds the sample to the experiemnt
+- calls `salmon._prepare_files()`, passing in  og_file_1
+- calls `check_salmon_quant` to run salmon quant and tximport
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
+    - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.
+- asserts that the experiment is not ready for tximport yet - salmon qunat still needs to be run on sample2
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
- asserts that the experiment is not ready for tximport yet - salmon quant still needs to be run on sample2
```

File: salmon.md [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+- asserts that there are no tximport outputs yet
+- calls `salmon._prepare_files()`, passing in  og_file_2
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
- calls `salmon._prepare_files()`, passing in og_file_2
```

File: salmon.md [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+- calls `prepare_organism_indices`
+    - creates a dummy OrganismIndex that can be used in the salmon pipeline
+- creates the test experiment
+- creates sample1 and og_file_1, associates the two, and adds the sample to the experiemnt
+- creates sample2 and og_file_2, associates the two, and adds the sample to the experiemnt
+- calls `salmon._prepare_files()`, passing in  og_file_1
+- calls `check_salmon_quant` to run salmon quant and tximport
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
+    - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

File: salmon.md [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+- asserts that the experiment is not ready for tximport yet - salmon qunat still needs to be run on sample2
+- asserts that there are no tximport outputs yet
+- calls `salmon._prepare_files()`, passing in  og_file_2
+- cleans up tximport data from previous runs if it exists
+- calls `check_salmon_quant` to run salmon quant and tximport
+    - asserts quant output file is made
+    - asserts quant output file strongly correlates to reference file
+    - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

File: salmon.md [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+    - at this point tximport should be run
+- asserts that the rds file (tximport output file) has been made
+- runs test_tximport.R to verify the output file
+- assert that two output files were created
+- assert that the two outut files exist
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
- assert that the two output files exist
```

File: salmon.md [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+
+### `test_get_tximport_inputs`
+
+Tests that tximport only considers RNA-Seq samples from GEO
```
Comment (jaclyn-taroni) [inline salmon.md]: I would expect this to be SRA not GEO? https://github.com/AlexsLemonade/refinebio-processor-documentation/pull/11/files#diff-76b3c308d473d4219815890b7c1da498a41bd8124fce60ff30938fc6dfe03e6eR21

File: salmon.md [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+    - By having 20/25 samples complete, we're just past both the numerical and percent cutoffs for running tximport early
+- calls `run_tximport_at_progress_point`
+    - creates a dummy `job_context` that could be passed into the salmon pipeline and have all steps related to tximport run successfully.
+    - creates a test experiment
+    - creates test samples and their associations for incomplete and complete samples and adds them to the test experiment
+    - runs tximport
+- asserts that 'tximported' is in the final `job_context`
+- for each sample in the completed samples list:
+    - check that the database associations were made properly 
+    - check that the tpm and rds files were created for the sample
```
Comment (jaclyn-taroni) [inline salmon.md]:
```suggestion
    - check that the tsv and rds files were created for the sample
```
I believe

**Review by jaclyn-taroni — APPROVED** [review]
Overall this looks really good! I only had minor comments.

File: salmon.md (hunk 16 above)
Comment (bug-sam) [reply salmon.md]: Yeah I think that's correct - this is incorrectly documented in the refinebio test file as well

**Pull request merged**